include/unifex/task.hpp

```diff
@@ -32,13 +35,171 @@
 namespace unifex {
 namespace _task {
 
+template <typename... Types>
+using _is_single_valued_tuple =
+    std::bool_constant<1 >= sizeof...(Types)>;
+
+template <typename... Types>
+using _is_single_valued_variant =
+    std::bool_constant<sizeof...(Types) == 1 && (Types::value &&...)>;
+
+template <typename Sender>
+UNIFEX_CONCEPT_FRAGMENT(     //
+  _single_typed_sender_impl, //
+    requires()(0) &&         //
+    sender_traits<remove_cvref_t<Sender>>
+      ::template value_types<
+        _is_single_valued_variant,
+        _is_single_valued_tuple>::value);
+
+template <typename Sender>
+UNIFEX_CONCEPT _single_typed_sender =
+  typed_sender<Sender> && UNIFEX_FRAGMENT(_single_typed_sender_impl, Sender);
+
+enum class state {empty, value, exception, done};
+
+template <typename T>
+struct _res {
+  struct type {
+    state state_ = state::empty;
+    union {
+      manual_lifetime<T> value_;
+      manual_lifetime<std::exception_ptr> exception_;
+    };
+    coro::coroutine_handle<> continuation_ {};
+    type() {}
+    ~type() {
+      switch(state_) {
+      case state::value:
+        value_.destruct();
+        break;
+      case state::exception:
+        exception_.destruct();
+        break;
+      default:;
+      }
+    }
+  };
+};
+template <typename T>
+using _result = typename _res<T>::type;
+
+template <typename T>
+struct _rec {
+  struct type {
+    _result<T> *res_;
+
+    template(class... Us)
+      (requires constructible_from<T, Us...> || //
+          (std::is_void_v<T> && sizeof...(Us) == 0))
+    void set_value(Us&&... us) && noexcept {
```

Comment:
Is that what you would prefer? I don't have strong feelings, except that whatever we decide, we should be consistent throughout unifex.

Reply:
I think that's what I've been doing in most algorithms...
I generally only use try/catch around parts that could fail after i've already destroyed the operation-state that invoked the set_value() method (eg. see `typed_via()`).

It would be interesting to get some data on codegen/performance of the different strategies though.